pdns/dnswriter.cc

```diff
@@ -183,145 +162,198 @@ void DNSPacketWriter::xfr8BitInt(uint8_t val)
-    if(ret->first == name)
-      break;
-  return ret;
-}
+  // iterate over the written labels, see if we find a match
+  const auto& raw = name.getStorage();
 
-// //! tokenize a label into parts, the parts describe a begin offset and an end offset
-// bool labeltokUnescape(labelparts_t& parts, const string& label)
-// {
-//   string::size_type epos = label.size(), lpos(0), pos;
-//   bool unescapedSomething = false;
-//   const char* ptr=label.c_str();
-
-//   parts.clear();
-
-//   for(pos = 0 ; pos < epos; ++pos) {
-//     if(ptr[pos]=='\\') {
-//       pos++;
-//       unescapedSomething = true;
-//       continue;
-//     }
-//     if(ptr[pos]=='.') {
-//       parts.push_back(make_pair(lpos, pos));
-//       lpos=pos+1;
-//     }
-//   }
-
-//   if(lpos < pos)
-//     parts.push_back(make_pair(lpos, pos));
-//   return unescapedSomething;
-// }
+  /* name might be a.root-servers.net, we need to be able to benefit from finding:
+     b.root-servers.net, or even:
+     b\xc0\x0c 
+  */
+  unsigned int bestpos=0;
+  *matchLen=0;
+  boost::container::static_vector<uint16_t, 34> nvect, pvect;
+
+  try {
+    for(auto riter= raw.cbegin(); riter < raw.cend(); ) {
+      if(!*riter)
+        break;
+      nvect.push_back(riter - raw.cbegin());
+      riter+=*riter+1;
+    }
+  }
+  catch(std::bad_alloc& ba) {
+    if(l_verbose)
+      cout<<"Domain "<<name<<" too large to compress"<<endl;
+    return 0;
+  }
+  
+  if(l_verbose) {
+    cout<<"Input vector for lookup "<<name<<": ";
+    for(const auto n : nvect) 
+      cout << n<<" ";
+    cout<<endl;
+    cout<<makeHexDump(string(raw.c_str(), raw.c_str()+raw.size()))<<endl;
+  }
 
+  if(l_verbose)
+    cout<<"Have "<<d_namepositions.size()<<" to ponder"<<endl;
+  int counter=1;
+  for(auto p : d_namepositions) {
+    if(l_verbose) {
+      cout<<"Pos: "<<p<<", "<<d_content.size()<<endl;
+      DNSName pname((const char*)&d_content[0], d_content.size(), p, true); // only for debugging
+      cout<<"Looking at '"<<pname<<"' in packet at position "<<p<<"/"<<d_content.size()<<", option "<<counter<<"/"<<d_namepositions.size()<<endl;
+      ++counter;
+    }
+    // memcmp here makes things _slower_
+    pvect.clear();
+    try {
+      for(auto iter = d_content.cbegin() + p; iter < d_content.cend();) {
+        uint8_t c=*iter;
+        if(l_verbose)
+          cout<<"Found label length: "<<(int)c<<endl;
+        if(c & 0xc0) {
+          uint16_t npos = 0x100*(c & (~0xc0)) + *++iter;
+          iter = d_content.begin() + npos;
+          if(l_verbose)
+            cout<<"Is compressed label to newpos "<<npos<<", going there"<<endl;
+          // check against going forward here
+          continue;
+        }
+        if(!c)
+          break;
+        pvect.push_back(iter - d_content.cbegin());
+        iter+=*iter+1;
+      }
+    }
+    catch(std::bad_alloc& ba) {
+      if(l_verbose)
+        cout<<"Domain "<<name<<" too large to compress"<<endl;
```

Comment:
If I'm not mistaken, `name` is not the culprit here, the message might be misleading.